crashtracker/src/collector/atomic_set.rs

```diff
@@ -0,0 +1,440 @@
+use rand::Rng;
+use std::fmt::Debug;
+use std::io::Write;
+use std::num::NonZeroU128;
+use std::ptr::null_mut;
+use std::sync::atomic::Ordering::SeqCst;
+
+pub(crate) type AtomicSpanSet<const LEN: usize> = AtomicSet<AtomicSpan, LEN>;
+pub(crate) type AtomicStringSet<const LEN: usize> = AtomicSet<AtomicString, LEN>;
+
+pub trait Atomic {
+    type Item: Ord + PartialEq + Debug + Clone;
+    const NONE: Self;
+    /// Returns whether there was a value before
+    fn clear(&self) -> bool {
+        self.take().is_some()
+    }
+    /// Returns whether there was anything to emit.
+    fn consume_and_emit(&self, w: &mut impl Write, leak: bool, first: bool)
+        -> anyhow::Result<bool>;
+    /// SAFETY: This is only safe to use in a single threaded context
+    #[cfg(test)]
+    unsafe fn load(&self) -> Option<Self::Item>;
+    /// Swaps the value with the old, returning the old
+    fn swap(&self, new: Option<Self::Item>) -> Option<Self::Item>;
+    /// Takes the value, leaving EMPTY_INNER in its place.
+    fn take(&self) -> Option<Self::Item> {
+        self.swap(None)
+    }
+    /// Returns `None` if the insert succeeded.
+    /// Returns `Some(val)` if the insert failed.
+    fn try_insert(&self, val: Self::Item) -> Option<Self::Item>;
+}
+
+pub struct AtomicSet<T, const LEN: usize> {
+    used: AtomicUsize,
+    set: [T; LEN],
+}
+
+impl<T, const LEN: usize> AtomicSet<T, LEN>
+where
+    T: Atomic,
+    <T as Atomic>::Item: std::cmp::PartialEq + Debug + Ord,
+{
+    /// Atomicity: This is NOT ATOMIC.  If other code modifies the set while this is happening,
+    /// badness will occur.
+    pub fn clear(&self) -> anyhow::Result<()> {
+        if !self.is_empty() {
+            for v in self.set.iter() {
+                if v.clear() {
+                    self.used.sub(1, SeqCst)
+                }
+            }
+        }
+        Ok(())
+    }
+
+    pub fn remove(&self, idx: usize) -> anyhow::Result<()> {
+        anyhow::ensure!(idx < self.set.len(), "Idx {idx} out of range");
+        if self.set[idx].clear() {
+            self.used.sub(1, SeqCst)
+        }
+        Ok(())
+    }
+
+    pub const fn new() -> Self {
+        Self {
+            used: AtomicUsize::new(0),
+            set: [T::NONE; LEN],
+        }
+    }
+
+    pub fn insert(&self, mut value: T::Item) -> anyhow::Result<usize> {
+        let used = self.used.fetch_add(1, SeqCst);
+        if used >= self.set.len() / 2 {
+            // We only fill to half full to get good amortized behaviour
+            self.used.fetch_sub(1, SeqCst);
+            anyhow::bail!("Crashtracker Atomic Set: No space to store {:?}", &value);
+        }
+
+        // Start at a random position.
+        // Since the array is only at most half full, and since we start scanning at random
+        // indicies, every slot should independently have <.5 probability of being occupied.
+        // Long scans become exponentially unlikely, giving amortized constant time insertion.
```

Comment:
This is linear probing, which should be OK given the high degree of entropy in the rng.  But I could implement a more clever probing strategy if we're worried about getting long linear chains.

Reply:
Changed to do n random scans, and only then a linear poll.